runtime/src/mm/cpp/Memory.cpp

```diff
@@ -0,0 +1,29 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+
+#include "Memory.h"
+
+#include "ThreadData.hpp"
+#include "ThreadRegistry.hpp"
+#include "Utils.h"
+
+extern "C" {
+
+struct MemoryState {
+    kotlin::mm::ThreadData data;
+
+    ALWAYS_INLINE static MemoryState* from(kotlin::mm::ThreadData* data) { return wrapper_cast(MemoryState, data, data); }
+};
+
+MemoryState* InitMemory() {
+    auto* data = kotlin::mm::ThreadRegistry::instance().RegisterCurrentThread();
+    return MemoryState::from(data);
```

Comment:
Somewhat fragile. It makes some sense to add static assertions checking there are no more fields in MemoryState.

Reply:
Left a single member inside `MemoryState` and left a comment explaining that it's a wrapper and asking not to add anything else to it. Should be sufficient.